### PR TITLE
Layout improvements

### PR DESCRIPTION
**Pull request by lkzhao**
* Add new "AxisDependentLayout" to provide helpers for layout that support both axis.
* Rework "CollectionLayout" interface to be more intuitive.
* Improve "FlexLayout" to support "justifyContent" & "alignItems"
* Make "FlowLayout" support horizontal axis

### CODE REVIEW

File: Sources/Layout/FlexLayout.swift [hunk 1]
```diff
@@ -21,22 +29,34 @@ public struct FlexValue {
...
 
-  public init(flex: [String: FlexValue], insets: UIEdgeInsets = .zero, padding: CGFloat = 0, axis: UILayoutConstraintAxis = .vertical) {
+  public init(
+    flex: [String: FlexValue] = [:],
+    insets: UIEdgeInsets = .zero,
+    padding: CGFloat = 0,
+    axis: Axis = .vertical,
+    justifyContent: FlexJustifyContent = .start,
+    alignItems: FlexAlignItem = .start
+  ) {
```
Comment (jindulys) [inline Sources/Layout/FlexLayout.swift]: nit: public init( 
        flex:...
        alignItems: FlexAlignItem = .start) {

File: Sources/Layout/AxisDependentLayout.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+public class AxisDependentLayout<Data>: CollectionLayout<Data> {
+  public var axis: Axis = .vertical
+
+  public override func doneLayout() {
+    visibleIndexSorter = axis == .vertical ? CollectionVerticalVisibleIndexSorter(frames: frames) : CollectionHorizontalVisibleIndexSorter(frames: frames)
+  }
+  
+  func primary(_ size: CGSize) -> CGFloat {
```
Comment (jindulys) [inline Sources/Layout/AxisDependentLayout.swift]: How about moving those method to a Util?

File: Sources/Layout/AxisDependentLayout.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+//  Copyright © 2017 lkzhao. All rights reserved.
+//
+
+import UIKit
+
+public enum Axis{
+  case vertical, horizontal
+}
+
+public class AxisDependentLayout<Data>: CollectionLayout<Data> {
```
Comment (jindulys) [inline Sources/Layout/AxisDependentLayout.swift]: Will a protocol help? This axisDependentLayout seems be able to use elsewhere

Comment (lkzhao) [inline Sources/Layout/AxisDependentLayout.swift]: I thought about protocol. but then every class have to store its own axis variable and also config their visibleIndexSorter according to their axis.

**Comment by jindulys** [comment]
+1, awesome! axis free!

**Pull request merged**